epub.css cleanup

poire-z opened the pull request:
Pinging @Frenzie @cramoisi @frankyifei for discussion.

I'd like to remove classname based selectors from epub.css, as it's the embedded stylesheet job to provide declarations for classes. See https://github.com/koreader/koreader/issues/3974#issuecomment-391824687 for an issue with what we currently have, and discussion in https://github.com/koreader/crengine/pull/191#issuecomment-391637844
We should still keep the few ones involving the `display:` property (because of the effect on the DOM with autoBoxes), and include them only for previously opened document depending on `gDOMVersionRequested`.

I just splitted epub.css (not changing any properties) into 3 sections (with a few indentation, spacing or `;` cleanups).

A first one, with element based declaration.

A second one below:
```
/* Old element or className base selectors involving display: that
 * we need to support for older gDOMVersionRequested */
```
A third one below:
```
/* Old element or className base selectors NOT involving display: that
 * we don't need to support even on previously opened documents */
```

I'd like to remove the third one (although it helps a bit on Wikipedia epub by making `<span class=citation>` italic, but I should then put that into the embedded stylesheet).

I'd need to find a way to include the second one only for older gDOMVersionRequested .

Not sure if I should do that on the lvstsheet.cpp side, with tricks like
`.code { -cr-gdomersionrequested-not-after: 20180526 ; display: block; }`
and just discard the whole declaration if we're higher than 20180526.

Or (easier) on the lua side with some CSS comments that would allow removing or allowing the section with some `gsub()` , but then the epub.css content would be read from lua and we won't be using anymore:
https://github.com/koreader/crengine/blob/e1a417f603fee02e14d812443f9f760b5f24934e/crengine/src/lvstsheet.cpp#L2816-L2840
but it would allow for a somehow cleaner:
```css
 dl { margin-left: 0em; } 
 dt { display: block; margin-left: 0em; margin-top:0.3em; font-weight: bold; } 
 dd { display: block; margin-left: 1.3em; } 

/* Old element or className base selectors involving display: that
 * we need to support for older gDOMVersionRequested */

/* ONLY-IF gDOMVersionRequested < 20180526
.title, .title1, .title2, .title3, .title4, .title5, .subtitle { 
    display: block; 
}
p image { display: inline }
.fb2_info { display: block; }
.code { display: block; }
ONLY-IF END */
```
(the `gsub` would remove the lines with ONLY-IF when an old document is opened)

Any thoughts?

poire-z commented:
Not so complicated to be able to handle:
```css
.title, .title1, .title2, .title3, .title4, .title5, .subtitle { 
    -cr-ignore-if-dom-version-greater-or-equal: 20180526;
    display: block; 
}
p image { -cr-ignore-if-dom-version-greater-or-equal: 20180526; display: inline }
```

cramoisi commented:
@poire-z in a more or less distant future (aka when everybody will have books opened after the updated dom version date), what will be easier to clean ?

poire-z commented:
Both should be easy to clean (remove declarations from epub.css, or remove the commented declaration).
The `-cr-ignore-if-dom-version-greater-or-equal: 20180526` that I just did just involves only crengine, so the frontend code is not polluted with these compatibility tricks.
So, I guess I made my choice :)

But do you agree on the principle of removing all these class name based declaration?
(I'll have to work on the wikipedia EPUB stylesheet.css to add some bit more then).

cramoisi commented:
@poire-z : I think I would agree on that even before the user side css set up, so now it's operational it make totally sense to do it. I prefer the -cr-ignore-if-dom-version-greater-or-equal: 20180526 too because it makes things less scattered (at least it seems to me) and more in their appropriate place

poire-z commented on cr3gui/data/epub.css:
```diff
@@ -1,94 +1,68 @@
-	text-indent: 0px; 
-	margin: 0; 
-	text-align: justify; 
+    text-indent: 0px; 
```
I didn't change any of the existing styles, so that I just don't have to wonder and check _what happens if_ :)
I was thinking that cleaning what remains would be a second step, for another PR.

Frenzie replied:
Fair enough.

cramoisi commented:
[not related]

@Frenzie Not that I botter as I compile by myself as I got a good setup (mainly thanks to you, guys) but there is no more update on http://build.koreader.rocks/download/nightly since 21/05, 1724

Frenzie commented:
The building is fine, but GitLab seems to have broken the mirrors.

poire-z commented:
OK, so I made the changes in small steps.
The idea was that no step update should bring any InfoMessage saying `Styles have changed in such a way that fully reloading the document may be needed for a correct rendering` for documents with a cache.
I somehow got some at the end with my last days documents :| It must have been because of some intermerdiate works.
I cleaned my cache, reverted to 20180524 version, opened a dozen of documents to make new caches, and tried each step on each, and didn't get any such message (I get one if I remove `small` from the` display: inline` list at the end, which tells me of the mistake).

So I guess this PR is fine and shouldn't break any past DOM/xpath as of step6.

One stylistic question: should we remove the lines with:
`display: block; /* default in fb2def.h, past and present */`
as it's the default (but I like some of them as they tell me on elements I wasn't really used to like UL and OL they are block, and LI only is the one with list-item).

More questionable is this about images:
```css
/* Images */
/* inline in fb2def.h, past and present */
img {
    text-align: center;
    text-indent: 0;
    display: block;
    /* should it really be block and centered when standalone?
     * or inline as img are everywhere else, and publisher would
     * have to wrap it in <div> to block/center it
     * Note that if block, lvrend.cpp displays the title="" content
     * under the image */
}
sup img { display: inline; }
sub img { display: inline; }
a img   { display: inline; }
p img   { display: inline; }
```
I would then need to add a` li img   { display: inline; }`, here or in wikipedia stylesheet.

Frenzie commented:
I think that images should be made block through publisher styles or surrounding elements like picture or div. Otherwise it's just… different from everywhere else.

cramoisi commented:
I'd say : should be made inline-block :)

poire-z commented:
No support for `inline-block` in crengine :) We're in nothing-fancy land.

Frenzie commented:
Does it even matter that much on images anyway? Most or possibly all inline elements can't have a width and height like an inline-block but images can.

poire-z commented:
Any comment about `nobr`? Is that some regular element?

poire-z commented:
OK, my cached test document survived this step 7.

(There's a thing when an image is the first element of a block node: the following text does not stick to it, as if there were a new line between them - this is not a style issue, but is to be fixed in lvrend.cpp/lvtextfm.cpp)

poire-z commented:
OK, removed my comment about `nobr` and bump CACHE_FILE_FORMAT_VERSION in a rebase of last commit.

cramoisi commented:
(I knew it but I was i joking mood)

poire-z commented:
So, OK with epub.css at the end of this PR?
Should I merge? (keep the multiple commits? or squash them into a single commit, and mention to go see this PR for the steps involved?)

Frenzie commented:
Either way is fine by me. My main complaint would be that the EOL stuff isn't combined with auto-format. Also I'm a bit confused about cleanup 4 counteracting the auto-format?

Frenzie commented on cr3gui/data/epub.css:
```diff
@@ -1,122 +1,170 @@
+/* With dom_version < 20180528, unknown elements defaulted to 'display: inherit'
+ * These ones here were explicitely set to inline (and some others not
+ * specified here were also set to inline in fb2def.h */
+b, strong, i, em, dfn, var, q, u, del, s, strike, small, big, sub, sup, acronym, tt, sa mp, kbd, code {
```
What's this sa mp thing?

poire-z replied:
A historical typo for `samp` ...  not to be fixed here as `samp` was then `display: inherit`. or just removed, but that would screw documents with `<sa><mp>strange</mp></sa>` :)

Frenzie replied:
Documents with what now? :-P

poire-z commented:
> My main complaint would be that the EOL stuff isn't combined with auto-format.

Yeah, I had to do the dos2unix first because `git diff` was showing some `^M` at the end of the lines I added (that github would not complain about).

> Also I'm a bit confused about cleanup 4 counteracting the auto-format?

and some more in step 7, as I like the single property declaration for the many inline elements each on a single line. I find it's easier to grasp the whole of them on 9 lines than on 27.

poire-z commented:
Mhhhh, given the pain of fixing that kind of stuff while ensuring backwards compatilibity for the sake of past highlights, I guess we'd better have a look at the other CSS files now, and never come back to that stuff.

The good news is that:
htm.css = rtf.css = txt.css , and nearly = chm.css and doc.css too (with only 1 margin-top and 1 margin-bottom for PRE less)
dict.css is ridiculously small, and I guess nobody ever read a .dict on koreader :)

So, that makes 4 different css:
- epub.css, that we just tweaked to be sane and readable
- fb2.css, with some specific FB2 stuff that we'd rather not touch as we don't really care
- htm.css (used for doc, rtf, txt, chm), which is based on FB2 (but the differences in formatting make a quick diff hard).
- and a cr3.css (symlink to it at the root of base/thirdparty/kpvcrlib) that credocument uses as a default for unknown extensions, and that seems to come from us (again, differences in formatting, based on FB2, and in a few % units that I managed to catch)...

It's bothering there are many similar, with the risk of diversion, multi editing, user believing that style may change anything and looking for changes...

May be we could just have epub.css be used for all document types, and fb2.css only for fb2 ?
And have readertypeset pick epub.css for all document formats except fb2?
That would clean this cr3gui/data, and our `Style>` submenu, which would look less rich.

I may look if there are any differences in display: between htm.css and epub.css, to see if we risk losing highlights. There probably are, and may be it would still need an individual htm.css with some -cr-ignore... stuff... Or we may just don't care about highlights in htm/rtf/doc/chm/txt and just go with epub.css?

Frenzie commented:
> dict.css is ridiculously small, and I guess nobody ever read a .dict on koreader :)

What is that exactly? Some kind of dictionary format I presume, but… :-P

> htm.css (used for doc, rtf, txt, chm), which is based on FB2 (but the differences in formatting make a quick diff hard).

Huh, I didn't realize that antiword had HTML output. But I definitely don't understand the split between HTML and EPUB, and possibly other HTML-based formats like CHM.

> May be we could just have epub.css be used for all document types, and fb2.css only for fb2 ?
And have readertypeset pick epub.css for all document formats except fb2?

Something along those lines definitely has my support. But what about those non-HTML formats like RTF?

cramoisi commented:
dict : https://en.wikipedia.org/wiki/DICT oO

who read pdb and chm, I wonder.

poire-z commented:
For reference, [htm_sorted.css.txt](https://github.com/koreader/crengine/files/2055330/htm_sorted.css.txt) is equivalent to htm.css and sorted and formatted following the same rules as the epub.css from this PR.
Just showing that there's nothing much lost by using epub.css instead of htm.css, except some FB2 tags at the end.
(And present in htm.css, that I should add to epub.css: `tt, samp, kbd` similar to `code`)

diff epub.css htm_sorted.css.txt:
```diff
--- epub.css
+++ htm_sorted.css
@@ -4,44 +4,41 @@
     text-align: justify;
 }

-head, script {
+head, form, script {
     display: none;
 }

-/* EPUB container of each individual HTML file */
-DocFragment {
-    page-break-before: always;
-}
-
 /* Headings */
 h1, h2, h3, h4, h5, h6 {
-    margin-top: 0.7em;
+    margin-top: 0.5em;
     margin-bottom: 0.5em;
     font-weight: bold;
     text-align: center;
     text-indent: 0;
     hyphenate: none;
-    adobe-hyphenate: none;
-}
-h1, h2, h3 {
-    page-break-before: always;
     page-break-inside: avoid;
     page-break-after: avoid;
+    display: block;
 }
-h4, h5, h6 {
-    page-break-inside: avoid;
-    page-break-after: avoid;
+
+h1, h2 {
+    page-break-before: always;
+    padding: 10px;
 }
+
 h1 { font-size: 150%; }
-h2 { font-size: 140%; }
-h3 { font-size: 130%; }
-h4 { font-size: 120%; }
-h5 { font-size: 110%; }
+h2 { font-size: 135%; }
+h3 { font-size: 125%; }
+h4 { font-size: 125%; font-style: italic; }
+h5 { font-size: 115%; font-style: italic; }
+h6 { font-size: 105%; font-style: italic; }

-/* Block elements */
-div {
-    margin: 1px;
+h1 p, h2 p, h3 p, h4 p, h5 p, h6 p {
+    text-align: center;
+    text-indent: 0;
 }
+
+/* Block elements */
 p {
     text-align: justify;
     text-indent: 1.2em;
@@ -84,8 +81,7 @@

 /* Tables */
 table {
-    font-size: 80%;
-    margin: 3px;
+    font-size: 100%;
 }
 td, th {
     text-indent: 0;
@@ -94,23 +90,21 @@
 th {
     font-weight: bold;
     text-align: center;
-    background-color: #DDD;
 }
 table caption {
     text-indent: 0;
     padding: 4px;
-    background-color: #EEE;
 }

 /* Monospace block and inline elements */
-pre, xmp {
+pre {
     white-space: pre;
     font-family: "Droid Sans Mono", "Liberation Mono", "DeJaVu Sans Mono", monospace;
     text-align: left;
     margin-top: 0.5em;
     margin-bottom: 0.5em;
-    /* background-color: #BFBFBF; */
 }
+
 code {
     white-space: pre;
     font-family: "Droid Sans Mono", "Liberation Mono", "DeJaVu Sans Mono", monospace;
@@ -122,10 +116,15 @@
 small                   { font-size: 80%; }
 big                     { font-size: 130%; }
 b, strong               { font-weight: bold; }
-i, em, dfn, var, cite   { font-style: italic; }
+i, em, dfn, var         { font-style: italic; }
 u                       { text-decoration: underline; }
-del, s, strike          { text-decoration: line-through; }
-a                       { text-decoration: underline; color: gray; }
+del, s, strike,strikethrough { text-decoration: line-through; }
+a                       { text-decoration: underline; }
+tt, samp, kbd           { font-family: "Liberation Mono", "DeJaVu Sans Mono", monospace; }
+
+b, strong, i, em, dfn, var, q, u, del, s, strike, small, big, sub, sup, acronym, tt, sa mp, kbd, code {
+    display: inline;
+}

 nobr {
     display: inline;
@@ -133,3 +132,166 @@
     white-space: nowrap;
 }

+img {
+    margin: 0.5em;
+    text-align: center;
+    text-indent: 0;
+    border-style: solid;
+    border-width: medium;
+}
+
+p image {
+    display: inline;
+}
+
+
+
+/************************ FB2 ******************/
+
+title p, subtitle p {
+    text-align: center;
+    text-indent: 0;
+}
+
+title {
+    font-size: 110%;
+    font-weight: bold;
+    text-align: center;
+    text-indent: 0;
+    hyphenate: none;
+    page-break-before: always;
+    page-break-inside: avoid;
+    page-break-after: avoid;
+    margin-top: 0.7em;
+    margin-bottom: 0.5em;
+}
+
+subtitle {
+    font-weight: bold;
+    font-style: italic;
+    text-align: center;
+    text-indent: 0;
+    hyphenate: none;
+    page-break-inside: avoid;
+    page-break-after: avoid;
+    margin-top: 0.3em;
+    margin-bottom: 0.3em;
+}
+
+cite p, epigraph p {
+    text-align: left;
+    text-indent: 0;
+}
+
+v {
+    text-align: left;
+    text-align-last: left;
+    text-align-last: left;
+    text-indent: 1em hanging;
+}
+
+annotation {
+    margin-left: 5%;
+    margin-right: 5%;
+    font-size: 100%;
+    font-style: italic;
+    text-align: justify;
+    text-indent: 1em;
+}
+
+epigraph {
+    margin-left: 15%;
+    margin-right: 1em;
+    text-align: left;
+    text-indent: 1em;
+    font-style: italic;
+    margin-top: 15px;
+    margin-bottom: 25px;
+}
+
+cite {
+    font-style: italic;
+    margin-left: 5%;
+    margin-right: 5%;
+    text-align: justify;
+    margin-top: 20px;
+    margin-bottom: 20px;
+}
+
+stanza + stanza {
+    margin-top: 1em;
+}
+
+stanza {
+    margin-left: 15%;
+    text-align: left;
+    font-style: italic;
+}
+
+poem {
+    margin-top: 1em;
+    margin-bottom: 1em;
+    text-indent: 0;
+}
+
+text-author {
+    font-weight: bold;
+    font-style: italic;
+    margin-left: 5%;
+}
+
+date {
+    display: block;
+    font-size: 80%;
+    font-style: italic;
+    text-align: center;
+}
+
+emphasis { font-style: italic; }
+
+empty-line { height: 1em; }
+
+description { display: block; }
+title-info { display: block; }
+genre { display: none; }
+author { display: none; }
+book-title { display: none; }
+keywords { display: none; }
+lang { display: none; }
+src-lang { display: none; }
+translator { display: none; }
+document-info { display: none; }
+publish-info { display: none; }
+custom-info { display: none; }
+coverpage { display: none; }
+
+
+image {
+    text-align: center;
+    text-indent: 0;
+    display: block;
+}
+
+a[type="note"] {
+    vertical-align: super;
+    font-size: 80%;
+    text-decoration: none;
+}
+
+body[name="notes"] {
+    font-size: 80%;
+}
+
+body[name="notes"] section title {
+    display: run-in;
+    text-align: left;
+    font-size: 110%;
+    font-weight: bold;
+    page-break-before: auto;
+    page-break-inside: auto;
+    page-break-after: auto;
+}
+
+body[name="notes"] section title p {
+    display: inline;
+}
+

```

poire-z commented:
Anyway, we can't really remove the other css, as these css may be requested from the sdr/metadata.lua, and if we were using something else, we may break past highlights (we could may be remove the duplicates and translate the removed to the kept one in readertypeset.lua, but that would complicate explaining the remap).

May be we could move (virtually, in readertypeset.lua) the others except epub.css and fb2.css into an `Style > Obsoleted >` submenu, and add a comment in caps at top to each obsolete cr3gui/data/*.css saying these are no more used and should not be modified?

Keep using the specified in metadata.lua even if obsoleted.
Only use epub.css (on all) and fb2.css (on .fb2 only) for new books.

Frenzie commented:
Remember when I mentioned @import? ;-)

poire-z commented:
Yes, but I get that that would just allows shortening the duplicates.

poire-z commented:
One thing that I kept and was introduced in 2011 by https://github.com/koreader/crengine/commit/3068274096c8ec9fb6d5cc8512317092c05d71e5 with no obvious reason or explanation:
https://github.com/koreader/crengine/blob/e562207a1ea80fcdb63c83347d540c15efc7a7fb/cr3gui/data/epub.css#L85-L89
Table then have a smaller font, and imbricated tables have a smaller than smaller font size and may become hard to read/tap on text.
I guess having initially a smaller font may make table look nicer/be more readable by having more text on in a possibly narrow column.
Should we remove this `font-size: 80%;` ?
Or use a trick like another css declaration like `table table {font-size: 100%} ` so imbricated tables don't get ridiculously small?

poire-z commented:
OK, having a smaller font size is better for table layout, and at the end for readability.
90% could be still OK, but as nobody complained about 80%, let's keep 80%.
I'll just add in #221:
```css
 /* Tables */
 table {
     font-size: 80%;
     margin: 3px;
 }
+table table { /* stop imbricated tables from getting smaller */
+    font-size: 100%;
+}
```
